internal/primitives/runtime/generic/header.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package generic
+
+import (
+	"io"
+
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+// Header is a block header, and implements a compatible encoding to `sp_runtime::generic::Header`
+type Header[N runtime.Number, H runtime.Hash, Hasher runtime.Hasher[H]] struct {
+	// The parent hash.
+	parentHash H
+	// The block number.
+	number N
+	// The state trie merkle root
+	stateRoot H
+	// The merkle root of the extrinsics.
+	extrinsicsRoot H
+	// A chain-specific digest of data useful for light clients or referencing auxiliary data.
+	digest runtime.Digest
+}
+
+// Number returns the block number.
+func (h Header[N, H, Hasher]) Number() N {
+	return h.number
+}
+
+// SetNumber sets the block number.
+func (h *Header[N, H, Hasher]) SetNumber(number N) {
+	h.number = number
+}
+
+// ExtrinsicsRoot returns the extrinsics root.
+func (h Header[N, H, Hasher]) ExtrinsicsRoot() H {
+	return h.extrinsicsRoot
+}
+
+// SetExtrinsicsRoot sets the extrinsics root.
+func (h *Header[N, H, Hasher]) SetExtrinsicsRoot(root H) {
+	h.extrinsicsRoot = root
+}
+
+// StateRoot returns the state root.
+func (h Header[N, H, Hasher]) StateRoot() H {
+	return h.stateRoot
+}
+
+// SetStateRoot sets the state root.
+func (h *Header[N, H, Hasher]) SetStateRoot(root H) {
+	h.stateRoot = root
+}
+
+// ParentHash returns the parent hash.
+func (h Header[N, H, Hasher]) ParentHash() H {
+	return h.parentHash
+}
+
+// SetParentHash sets the parent hash.
+func (h *Header[N, H, Hasher]) SetParentHash(hash H) {
+	h.parentHash = hash
+}
+
+// Digest returns the digest.
+func (h Header[N, H, Hasher]) Digest() runtime.Digest {
+	return h.digest
+}
+
+// DigestMut returns a mutable reference to the stored digest.
+func (h Header[N, H, Hasher]) DigestMut() *runtime.Digest {
```

Comment:
I do not see the need for this function, and if we do need we should rename it to better fit go

Reply:
see https://github.com/ChainSafe/gossamer/pull/4015#discussion_r1644924606